File: code/java/63.java
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int row = obstacleGrid.length;
        int col = obstacleGrid[0].length;
        int[][] dp = new int[row][col];

        dp[0][0] = 1;
        for(int i=1;i<row;i++) dp[i][0] = (obstacleGrid[i][0]==1||dp[i-1][0]==0)?0:1;
        for(int i=1;i<col;i++) dp[0][i] = (obstacleGrid[0][i]==1||dp[0][i-1]==0)?0:1;

        for(int i=1;i<row;i++){
            for(int j=1;j<col;j++){
                dp[i][j] = obstacleGrid[i][j]==1?0:dp[i-1][j]+dp[i][j-1];
            }
        }
        return dp[row-1][col-1];
    }
}

File: code/python/63.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        if obstacleGrid[0][0]==1:
            return 0
        row = len(obstacleGrid)
        col = len(obstacleGrid[0])
        dp = [[0]*col for _ in range(row)]
        for i in range(row):
            if obstacleGrid[i][0]==0:
                dp[i][0] = 1
            else:
                break
        for i in range(col):
            if obstacleGrid[0][i]==0:
                dp[0][i] = 1
            else:
                break
        for i in range(1,row):
            for j in range(1,col):
                if obstacleGrid[i][j]==0:
                    dp[i][j] = dp[i-1][j]+dp[i][j-1]
        #print(dp)
        return dp[-1][-1]